### DISCUSSION TITLE
Naming convention for setter and getter

### DISCUSSION THREAD

**Discussion by ruv** [ForthHub/discussion#137] [Show and tell]
## `set-*`, not `get-*`

Let's say we have two separate words, one plays the role of a getter and produces _x.foo_ as `( -- x.foo )`, and another plays the role of a setter and consumes _x.foo_ as `( x.foo -- )`, and they both are bound to some object "foo". It could seem worthwhile to name these words as `get-foo` and `set-foo`.

But this choice is not very good. The names "**s**et-foo" and "**g**et-foo" are distinct only in one letter, while the words are distinct drastically by semantics (it [was pointed out](https://groups.google.com/g/comp.lang.forth/c/eJ5sH-2IFUM/m/6L_slB6Qi3wJ) by Michael Gassanenko many years ago).

A better variant is to use just `foo` for a getter, and `set-foo` for a setter.

Getters are used far more frequently than setters, so it's good if their name is shorter.
In its turn, setters should be **more prominent**, since they change a state. A longer name meets this goal.

An example of a standard words pair that follow this convention is [`precision`](https://forth-standard.org/standard/float/PRECISION) and [`set-precision`](https://forth-standard.org/standard/float/SET-PRECISION).

Sometimes a setter can be named as `!foo`, `->foo`, or `>>foo` (for example, see a kind of [named parameters in Factor](https://concatenative.org/wiki/view/Concatenative%20language/Keyword%20parameters)).
Though, source code with many special symbols is worse to read. Therefore, names without special symbols is a better choice for arbitrary open APIs.

## Why not `foo!` for setters

The standard words that end with "`!`" take an address/object as the top argument:

- `defer! ( xt xt.defer -- )`
- `c! ( char c-addr -- )`
- `2! ( 2*x a-addr -- )`
- `f! ( F: r  S: f-addr -- )`

The general form for them is:
- `bar! ( bar-data addr -- )`
(where _bar-data_ in the stack diagram is the data type of a consumed data object)

Obviously, this convention is not suitable for a setter `( foo-data -- )`.  Therefore, `!foo` is a better choice than `foo!` for a setter.

## Neither `foo@` nor `@foo` for getters

Similarly, variant `foo@` is a bad choice for a getter `( -- foo-data )`. It does not fit the convention of the standard words that end with "`@`" and read an address/object:

- `defer@ ( xt.defer -- xt )`
- `c@ ( c-addr -- char )`
- `2@ ( a-addr -- 2*x )`
- `f@ ( S: f-addr --  F: -- r )`

And the variant `@foo` is worse than just `foo` for a getter — since it's longer, and it contains a special symbol without enough ground.

## Other prefixes

Depending on particular use cases, other prefixes can be employed — to encode additional information into getter/setter names. For example, `is-`, or `to-`.

This is similar to how `to` is used for VALUEs, and `is` is used for DEFERs — such separation provides more useful information.

**Comment by SirWumpus** [comment]
I've always had a preference for:

* `get_foo`
* `put_foo` which has at least 2 distinct letters from `get_foo`.
* `is_foo` as a question that returns a boolean; some variants have been `foo?` or `?foo`.

While I appreciate the argument for shorter getter name, I prefer the symmetry of `put_foo` particularly when reading the code more like a sentence.  Its clear what the objective is. 

(I prefer snake case `_` over `-`, mainly because of how many editors handle selection of identifiers / words.  But that's a different discussion.)

**Reply by CCurl** [reply]
I do this:

variable (foo)
: foo (foo) @ ;
: foo! (foo) ! ;

It works for me.

**Reply by ruv** [reply]
@SirWumpus writes:

> `put_foo` which has at least 2 distinct letters from `get_foo`

To me, `put` is associated with output — printing or sending.

One more argument against `get-foo` — `get` is a verb, so the whole name becomes a verb.

> `is_foo` as a question that returns a boolean

It was tempting for me too.  But the pattern `['] x is y` is used in Forth, and `is` does not imply a question here.
Then, `['] x is-y` should not imply a question too.  In English, `is` implies a question in the beginning of a (sub-) sentence only.

> some variants have been `foo?` or `?foo`.

Regarding `?foo` — I will write a separate post.

Regarding `foo?` — for a word that return a flag I try to find such a name that already tells you that it's a boolean, without a question mark.

My recent finding is a _plumbing_ API to access and change the Forth text interpreter state (instead of `STATE` and as a basis for `[` and `]`):
- `enter-compilation ( -- )` — enter compilation state (if it isn't entered yet, otherwise do nothing)
- `leave-compilation ( -- )` — leave compilation state (if it's entered, otherwise do nothing), i.e., enter interpretation state.
- `compilation ( -- flag )`  — return a flag whether the system in compilation state; e.g.: `compilation if do-this else to-that then`

It's clear that `compilation` just return something since **nouns** should never have side effects (change state, etc).


> I prefer snake case `_` over `-`, mainly because of how many editors handle selection of identifiers / words. 

Many languages allow a dash in identifiers/names. I would better change my editor than abandon dashes 😂 


@CCurl writes:
> I do this:
> ```
> variable (foo)
> : foo (foo) @ ;
> : foo! (foo) ! ;
> ``` 
> It works for me.

I used to use the form `foo!` for setters in past, but changed my mind under pressure of the arguments 😄

When you mix this form for setters and for words that write something into a given location/object, a fragment of source code like `bar baz foo!` just brings you less information.

**Reply by CCurl** [reply]
Of course "boo baz foo!" means nothing. A more meaningful example would be:

variable (rows)
: rows (rows) @ ; : rows! (rows) ! ;

variable (cols)
: cols (cols) @ ; : cols! (cols) ! ;

100 rows! 50 cols!

... rows 0 do cols 0 do ... loop loop ...

It's natural and "intuitively obvious upon casual inspection" to me that "rows" gives me the number of rows and "rows!" sets the number of rows.

**Reply by ruv** [reply]
@CCurl writes:
> `100 rows! 50 cols!`

Well, `100 !rows 50 !cols` is also "intuitively obvious upon casual inspection".
Moreover, the form `!foo` is even more prominent than `foo!`.

A general idea is to **use different naming forms for different style of words**, when it's possible.

And since the form `foo!` is already used for the standard words having style `( data object -- )`, then it's better to use another form for the words having style `( data -- )`, which are bound to an object and change its state.

**Reply by CCurl** [reply]
Sorry ... I am not going to change the programming paradigm I prefer simply because someone on the internet says that form is "already used".

That is supposed to be the strength of Forth; that the developer has the freedom to define the syntax that works for him.

Why try to take that freedom away?

Sure, for a team that is working together on a project, some programming style standards would probably be a good idea. Then I could see a discussion of this type of thing in that context, but even then, it's 100% up to the team to decide what works for them, not us.

**Reply by ruv** [reply]
> because someone on the internet says that form is "already used"

It's already used by you too, in the words like `pina!`, `pin!`, `c!`, `w!`, etc. 

Of course, it's up to the reader whether it is worth changing their practice or take these ideas into account in a new project.


> Why try to take that freedom away?

Nobody tries to take that freedom away. It's merely words on the screen 😉 

I only provide reasoning or **arguments** why one choice is better or worse than another. If you disagree, you can provide counter-arguments and we are free to talk about that. It's also fine to just ignore them.

**Reply by CCurl** [reply]
Fair enough. You have your preferences, I have mine. Let's leave it at that and move on.
